### ISSUE TITLE
Training becomes NaN around epoch 8–12 under the paper's reported settings

### ISSUE THREAD

**Issue by hplikespepper** [MKnoche/DONUT#4]
Hi, thank you for releasing this work and the code, it's a really good work.

I tried to train the model using the same settings reported in the paper (4 gpus, same lr, and batch size etc.). At the beginning, training seems normal and everything looks fine. However, in every run, around epoch 8–12, all values shown in the tqdm bar become **NaN**.

I also checked the recorded log `metrics.csv`, and after that point there are no valid numbers anymore — all metrics are `NaN`, like this:
"9,30549,,,,nan,,nan,,,,,nan,nan,,,nan,nan,,,,,,,,,,"

I would like to ask:
  * Have you encountered this issue during training?
  * Is there any known cause (e.g., learning rate, mixed precision, gradient explosion, data issue, etc.)?
  * Are there any recommended settings or fixes to avoid this?
Thanks in advance for your help.

**Comment by MKnoche** [comment]
Thanks for you interest in our work and for the kind words!

The published code is a cleaned-up version from the original research code, but I have since used this released codebase as the basis for my current research without observing NaNs so far.

Did you start all the different runs from scratch, or did they auto-resume from a checkpoint? Maybe the model got into a bad state at some point and could not recover from that.

**Comment by hplikespepper** [comment]
Hi, thank you so much for your reply. I started all the different runs from scratch, I used the parameter settings from the paper, with the only difference being that I used 4 H200 GPUs.
Since the training runs normally on your machine, the issue is likely on my side. I will check it more carefully. Thanks again.

**Comment by binbin0801** [comment]
Hi, I encountered a similar problem. Have you solved it?

**Comment by hplikespepper** [comment]
> Hi, I encountered a similar problem. Have you solved it?

In modules/donut_layer.py, replace line 172-173 with:
        xs['scale'] = (1 + elu(xs['scale'])).clamp(min=1e-4)
        xs['conc'] = (1 + elu(xs['conc'])).clamp(min=1e-4)
Also, in predictors/donut.py:
replace line 254 with "pred[0]['conc'].append(1/xs['conc'].clamp(min=1e-4))", 
replace line 274 with "pred[1]['conc'].append(1/xs['conc'].clamp(min=1e-4))".

The issue is triggered by the value of "xs['conc']" (thanks to Claude, manually checking for this issue would be completely unrealistic.).

**Comment by binbin0801** [comment]
> > Hi, I encountered a similar problem. Have you solved it?
> 
> In modules/donut_layer.py, replace line 172-173 with: xs['scale'] = (1 + elu(xs['scale'])).clamp(min=1e-4) xs['conc'] = (1 + elu(xs['conc'])).clamp(min=1e-4) Also, in predictors/donut.py: replace line 254 with "pred[0]['conc'].append(1/xs['conc'].clamp(min=1e-4))", replace line 274 with "pred[1]['conc'].append(1/xs['conc'].clamp(min=1e-4))".
> 
> The issue is triggered by the value of "xs['conc']" (thanks to Claude, manually checking for this issue would be completely unrealistic.).

Thanks for sharing the solution!

**Comment by MKnoche** [comment]
Thank you! I have added the clamping to `modules/donut_layer.py` (962c05a15a82bff5ea47541e9fb3842c782b2a1c). The change in `predictors/donut.py` would be redundant as it is would just clamp the same tensor again.

I also double‑checked this part in the original training code; it is identical to the released version, and I never observed this issue on my setup. Maybe it depends on the specific hardware or some GPU settings. In any case, clamping should not hurt.
